### PR TITLE
More precise parsing

### PR DESCRIPTION
**Pull request by tclem**
Enables precise ASTs for:

- [x] Go
- [x] TypeScript
- [x] TSX (and JSX and JavaScript)

### CODE REVIEW

File: semantic-go/src/Language/Go/Tags.hs [hunk 1]
```diff
@@ -0,0 +1,95 @@
+{-# LANGUAGE AllowAmbiguousTypes, DataKinds, DisambiguateRecordFields, FlexibleContexts, FlexibleInstances, MultiParamTypeClasses, NamedFieldPuns, PartialTypeSignatures, ScopedTypeVariables, TypeApplications, TypeFamilies, TypeOperators, UndecidableInstances #-}
```
Comment (aymannadeem) [inline semantic-go/src/Language/Go/Tags.hs]: can we use the alternate style with one pragma per extension, each on their own line?

Comment (tclem) [reply semantic-go/src/Language/Go/Tags.hs]: Yes, that's a very good idea. I'll fix that for at least the new files I've added on this PR. 👍

File: semantic-tsx/src/Language/TSX.hs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{-# OPTIONS_GHC -freduction-depth=0 #-}
```
Comment (tclem) [inline semantic-tsx/src/Language/TSX.hs]: @robrix - OK to leave this in for now? Or...

File: test/Examples.hs [hunk 3]
```diff
@@ -78,17 +103,51 @@ rubySkips = Path.relFile <$>
...
 
   -- Can't detect method calls inside heredoc bodies with precise ASTs
   , "ruby_spec/core/argf/readpartial_spec.rb"
   , "ruby_spec/core/process/exec_spec.rb"
   ]
 
+tsxSkips :: [Path.RelFile]
+tsxSkips = Path.relFile <$>
+  [
+  -- Cannot decode byte '\xe2': Data.Text.Internal.Encoding.decodeUtf8: Invalid UTF-8 stream
```
Comment (tclem) [inline test/Examples.hs]: Almost all languages have this same kind of encoding problem with certain example files that contain non-utf8 characters. Should we just leave this as a known issue for now?

File: test/fixtures/go/corpus/assignment-statements.diffA-B.txt [hunk 4]
```diff
@@ -8,12 +8,7 @@
       { (Identifier)
       ->(Identifier) }
         (Integer))
-      (Assignment
-        (Statements
-        { (Identifier)
-        ->(Identifier) }
-        { (Identifier)
-        ->(Identifier) })
+      (AugmentedAssignment
```
Comment (tclem) [inline test/fixtures/go/corpus/assignment-statements.diffA-B.txt]: Various changes to shape of assignment trees due to things like adding `AugmentedAssignment`.

File: semantic-tsx/src/Language/TSX.hs (hunk 2 above)
Comment (robrix) [reply semantic-tsx/src/Language/TSX.hs]: 👍

File: test/Examples.hs (hunk 3 above)
Comment (robrix) [reply test/Examples.hs]: cc @maxbrunsfeld

Comment (maxbrunsfeld) [reply test/Examples.hs]: What is the error? Tree-sitter *should* just treat this as a syntax error; it shouldn't halt the whole thing.

Comment (tclem) [reply test/Examples.hs]: `Cannot decode byte '\xe2': Data.Text.Internal.Encoding.decodeUtf8: Invalid UTF-8 stream` is the error, it's coming from [unmarshalling](https://github.com/tree-sitter/haskell-tree-sitter/blob/master/tree-sitter/src/TreeSitter/Unmarshal.hs#L124) in haskell-tree-sitter.

Comment (maxbrunsfeld) [reply test/Examples.hs]: Oh, I see. @robrix @aymannadeem For that Text-unmarshalling step, can we do some kind of *lossy* UTF8 decoding, e.g. in which invalid byte sequences are replaced by the [replacement character](https://en.wikipedia.org/wiki/Specials_(Unicode_block)#Replacement_character). I don't think we want to bail out of analyzing the whole file if there are some non-utf8 bytes somewhere.

Comment (tclem) [reply test/Examples.hs]: Yes!

``` haskell
decodeUtf8With lenientDecode
```

Comment (robrix) [reply test/Examples.hs]: Oooh, my bad, I misread the error. Fortunately @tclem fixed this yesterday!

**Pull request merged**